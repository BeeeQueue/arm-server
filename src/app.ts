import process from "node:process"

import Fastify from "fastify"
import { customAlphabet, urlAlphabet } from "nanoid"

import Cors from "@fastify/cors"
import Helmet from "@fastify/helmet"

import pkgJson from "../package.json"
import { config } from "@/config"
import { docsPlugin } from "@/docs"
import { logger } from "@/lib/logger"
import { sendErrorToSentry } from "@/lib/sentry"
import { apiPlugin } from "@/routes/v1/ids/handler"
import { v2Plugin } from "@/routes/v2/ids/handler"
import { thetvdbPlugin } from "@/routes/v2/thetvdb/handler"
import { CacheTimes, cacheReply } from "@/utils"

const PROD = config.NODE_ENV === "production"

const nanoid = customAlphabet(urlAlphabet, 16)

export const buildApp = async () => {
  const App = Fastify({
    ignoreTrailingSlash: true,
    onProtoPoisoning: "remove",
    onConstructorPoisoning: "remove",
    trustProxy: PROD,
    genReqId: () => nanoid(),
    disableRequestLogging: process.env.NODE_ENV === "test",
    logger,
  })

  await App.register(Cors, {
    origin: true,
  })

  await App.register(Helmet, {
    hsts: false,
    contentSecurityPolicy: false,
  })

  App.addHook("onError", (request, reply, error, next) => {
    /* c8 ignore next 4 */
    if (error.validation != null) {
      if (request.method === "GET") {
        cacheReply(reply, CacheTimes.WEEK)
      }
    } else {
      sendErrorToSentry(error, request)
    }

    next()
  })

  await App.register(apiPlugin, { prefix: "/api" })
  await App.register(v2Plugin, { prefix: "/api/v2" })
  await App.register(thetvdbPlugin, { prefix: "/api/v2" })
  await App.register(docsPlugin)

  App.get("/", async (_, reply) => {
    cacheReply(reply, CacheTimes.WEEK * 2)
    void reply.redirect(301, pkgJson.homepage)
  })

  return App
}
