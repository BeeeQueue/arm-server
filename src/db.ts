import Knex from 'knex'
import { config } from '../knexfile'

const { NODE_ENV } = process.env
export const knex = Knex(
  config[NODE_ENV as 'development' | 'production' | 'test']
)

export interface Relation {
  anilist?: number
  anidb?: number
  myanimelist?: number
  kitsu?: number
}
