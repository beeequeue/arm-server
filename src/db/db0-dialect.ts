/* eslint-disable ts/consistent-type-imports */

import type { Connector, Database, Primitive } from "db0"
import {
	type CompiledQuery,
	type DatabaseConnection,
	type DatabaseIntrospector,
	type Dialect,
	type Driver,
	type Kysely,
	type QueryCompiler,
	type QueryResult,
	SqliteAdapter,
	SqliteIntrospector,
	SqliteQueryCompiler,
} from "kysely"

type SqliteDatabase = import("node:sqlite").DatabaseSync
// | import("better-sqlite3").DatabaseSync
// | import("sqlite3").DatabaseSync
// | import("@cloudflare/workers-types/latest").D1Database

/**
 * todo docs
 */
export class Db0SqliteDialect implements Dialect {
	#db: Database<Connector<SqliteDatabase>>

	constructor(db: Database<Connector<SqliteDatabase>>) {
		this.#db = db
	}

	createAdapter() {
		return new SqliteAdapter()
	}

	createDriver(): Driver {
		return new Db0SqliteDriver(this.#db)
	}

	createQueryCompiler(): QueryCompiler {
		return new SqliteQueryCompiler()
	}

	createIntrospector(db: Kysely<any>): DatabaseIntrospector {
		return new SqliteIntrospector(db)
	}
}

class Db0SqliteDriver implements Driver {
	#db: Database<Connector<SqliteDatabase>>

	constructor(db: Database<Connector<SqliteDatabase>>) {
		this.#db = db
	}

	async init(): Promise<void> {}

	async acquireConnection(): Promise<DatabaseConnection> {
		return new Db0SqliteConnection(this.#db)
	}

	async beginTransaction(conn: Db0SqliteConnection): Promise<void> {
		return conn.beginTransaction()
	}

	async commitTransaction(conn: Db0SqliteConnection): Promise<void> {
		return conn.commitTransaction()
	}

	async rollbackTransaction(conn: Db0SqliteConnection): Promise<void> {
		return conn.rollbackTransaction()
	}

	async releaseConnection(_conn: Db0SqliteConnection): Promise<void> {}

	async destroy(): Promise<void> {
		const instance = await this.#db.getInstance()

		// eslint-disable-next-line ts/no-unsafe-member-access
		if (typeof (instance as any).close === "function") {
			return (instance as never as { close: () => Promise<void> }).close()
		}
		// eslint-disable-next-line ts/no-unsafe-member-access
		if (typeof (instance as any).destroy === "function") {
			return (instance as never as { destroy: () => Promise<void> }).destroy()
		}
	}
}

class Db0SqliteConnection implements DatabaseConnection {
	#db: Database<Connector<SqliteDatabase>>

	// TODO: transactions?

	constructor(db: Database<Connector<SqliteDatabase>>) {
		if (db.dialect !== "sqlite") {
			throw new Error("db0 driver only supports sqlite dialects for now.")
		}

		this.#db = db
	}

	async executeQuery<O>(compiledQuery: CompiledQuery): Promise<QueryResult<O>> {
		// Transactions are not supported yet.
		// if (this.#transactionClient) return this.#transactionClient.executeQuery(compiledQuery)

		const results = await this.#db
			.prepare(compiledQuery.sql)
			.bind(...(compiledQuery.parameters as Primitive[]))
			.all()
			.then((data) => ({ data, error: null }))
			.catch((error: Error) => ({ data: null, error }))
		if (results.error) {
			console.dir(results.error)
			throw new Error("Failed to execute query", { cause: results.error })
		}

		const numAffectedRows =
			results.data.length > 0 ? BigInt(results.data.length) : undefined

		return {
			// insertId:
			//   results.meta.last_row_id === undefined || results.meta.last_row_id === null
			//     ? undefined
			//     : BigInt(results.meta.last_row_id),
			rows: (results.data as O[]) ?? [],
			numAffectedRows,
		}
	}

	async beginTransaction() {
		// this.#transactionClient = this.#transactionClient ?? new PlanetScaleConnection(this.#config)
		// this.#transactionClient.#conn.execute('BEGIN')
		throw new Error("db0 driver does not support transactions yet.")
	}

	async commitTransaction() {
		// if (!this.#transactionClient) throw new Error('No transaction to commit')
		// this.#transactionClient.#conn.execute('COMMIT')
		// this.#transactionClient = undefined
		throw new Error("db0 driver does not support transactions yet.")
	}

	async rollbackTransaction() {
		// if (!this.#transactionClient) throw new Error('No transaction to rollback')
		// this.#transactionClient.#conn.execute('ROLLBACK')
		// this.#transactionClient = undefined
		throw new Error("db0 driver does not support transactions yet.")
	}

	async *streamQuery<O>(
		_compiledQuery: CompiledQuery,
		_chunkSize: number,
	): AsyncIterableIterator<QueryResult<O>> {
		throw new Error("db0 driver does not support streaming (yet?)")
	}
}
