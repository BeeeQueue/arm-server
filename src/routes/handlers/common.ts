import Joi from 'joi'

import { enumToArray } from '@/utils'

export enum Source {
  AniList = 'anilist',
  AniDB = 'anidb',
  MAL = 'myanimelist',
  Kitsu = 'kitsu',
}

export const sourceArray = (enumToArray(Source) as unknown) as string[]

export const idSchema = Joi.number()
  .min(0)
  .max(2147483647)
  .precision(0)
  .required()
