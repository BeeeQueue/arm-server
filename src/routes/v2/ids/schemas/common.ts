import { z } from "zod"

// Does not include `thetvdb` due to the one-to-many issue
export const numberIdSourceSchema = z.enum([
	"anilist",
	"anidb",
	"anisearch",
	"kitsu",
	"livechart",
	"themoviedb",
	"myanimelist",
])

export const stringIdSourceSchema = z.enum(["anime-planet", "notify-moe"])
export const imdbSourceSchema = z.enum(["imdb"])
