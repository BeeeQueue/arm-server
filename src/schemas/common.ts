import { createSchema as S } from "ts-json-validator"

import { enumToArray } from "@/utils"

export enum Source {
  AniList = "anilist",
  AniDB = "anidb",
  MAL = "myanimelist",
  Kitsu = "kitsu",
}

export const sourceSchema = S({
  type: "string",
  enum: enumToArray(Source),
})

export const idSchema = S({
  type: "integer",
  minimum: 0,
  maximum: 50_000_000,
})
