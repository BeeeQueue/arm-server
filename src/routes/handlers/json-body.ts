import Joi from 'joi'

import { idSchema, Source, sourceArray } from '@/routes/handlers/common'
import { knex, Relation } from '@/db'

type BodyItem = {
  [key in Source]?: number
}

const bodyItemSchema = Joi.object(
  sourceArray.reduce(
    (obj, source) => ({
      ...obj,
      [source as any]: idSchema.optional(),
    }),
    {},
  ),
)
  .min(1)
  .required()

export type BodyQuery = BodyItem | BodyItem[]

const arraySchema = Joi.array().min(1).max(100).items(bodyItemSchema).required()

export const bodySchema = Joi.alternatives(arraySchema, bodyItemSchema)

export const bodyHandler = async (
  input: BodyQuery,
): Promise<
  BodyQuery extends Array<undefined> ? Array<Relation | null> : Relation
> => {
  if (!Array.isArray(input)) {
    const relation = await knex.where(input).from('relations').first()

    console.log(relation)

    return relation ?? null
  }

  let relations: Array<Relation | null> = ([] = [])

  // Get relations
  relations = await knex
    .where(function () {
      input.forEach((item) => this.orWhere(item))
    })
    .from('relations')

  // Map them against the input, so we get results like [{item}, null, {item}]
  relations = input.map((item) => {
    const realItem = Object.entries(item)[0] as [Source, number]

    return (
      relations.find((relation) => relation![realItem[0]] === realItem[1]) ??
      null
    )
  })

  return relations as any
}
