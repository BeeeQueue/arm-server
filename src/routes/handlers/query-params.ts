import Joi from 'joi'

import { idSchema, Source, sourceArray } from '@/routes/handlers/common'
import { knex } from '@/db'

export type QueryParamQuery = {
  source: Source
  id: number
}

export const querySchema = Joi.object({
  source: Joi.string()
    .valid(...sourceArray)
    .required(),
  id: idSchema,
})

export const handleQueryParams = async (input: QueryParamQuery) => {
  return knex
    .where({ [input.source]: input.id })
    .from('relations')
    .first()
}
