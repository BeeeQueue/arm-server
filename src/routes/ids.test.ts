import request from 'supertest'

import { App } from '@/app'
import { knex, Relation } from '@/db'
import { Source } from '@/routes/handlers/common'

let id = 0
const createRelations = async <N extends number>(
  amount: N,
): Promise<N extends 1 ? Relation : Relation[]> => {
  const relations: Relation[] = Array.from({ length: amount }).map(() => {
    const r = {
      anilist: id,
      anidb: id + 1,
      kitsu: id + 2,
      myanimelist: id + 3,
    }

    id += 4

    return r
  })

  await knex.insert(relations).into('relations')

  if (amount === 1) {
    return relations[0] as any
  }

  return relations as any
}

const server = App.listen()

afterAll(() => {
  server.close()
})

afterEach(() => knex.delete().from('relations'))

describe('query params', () => {
  test('fetches relation correctly', async () => {
    const relation = await createRelations(1)

    return request(server)
      .get('/api/ids')
      .query({
        source: Source.AniList,
        id: relation.anilist,
      })
      .expect(200, relation)
      .expect('Content-Type', /json/)
  })

  test("returns null when id doesn't exist", async () =>
    request(server)
      .get('/api/ids')
      .query({
        source: Source.Kitsu,
        id: 404,
      })
      .expect(200, null))
})

describe('json body', () => {
  describe('object input', () => {
    test('fetches a single relation', async () => {
      const relations = await createRelations(4)

      const body = {
        [Source.AniDB]: relations[0].anidb,
      }

      return request(server)
        .post('/api/ids')
        .send(body)
        .expect(relations[0])
        .expect('Content-Type', /json/)
    })

    test('errors correctly on an empty object', async () => {
      await createRelations(4)

      const response = {
        code: 400,
        error: 'Bad Request',
        validation: '"value" must have at least 1 key',
      }

      return request(server)
        .post('/api/ids')
        .send({})
        .expect(400, response)
        .expect('Content-Type', /json/)
    })

    test('returns null if not found', async () => {
      await createRelations(4)

      return request(server)
        .post('/api/ids')
        .send({ anidb: 100_000 })
        .expect(200, null)
        .expect('Content-Type', /json/)
    })
  })

  describe('array input', () => {
    test('fetches relations correctly', async () => {
      const relations = await createRelations(4)

      const body = [
        { [Source.AniDB]: relations[0].anidb },
        { [Source.AniList]: 1000 },
        { [Source.Kitsu]: relations[2].kitsu },
      ]

      const result = [relations[0], null, relations[2]]

      return request(server)
        .post('/api/ids')
        .send(body)
        .expect(result)
        .expect('Content-Type', /json/)
    })

    test('responds correctly on no finds', async () => {
      const body = [{ [Source.AniList]: 1000 }, { [Source.Kitsu]: 1000 }]

      const result = [null, null]

      return request(server)
        .post('/api/ids')
        .send(body)
        .expect(result)
        .expect('Content-Type', /json/)
    })

    test('requires at least one source', async () => {
      const body = [{}]

      return request(server)
        .post('/api/ids')
        .send(body)
        .expect(400)
        .expect('Content-Type', /json/)
    })
  })
})
