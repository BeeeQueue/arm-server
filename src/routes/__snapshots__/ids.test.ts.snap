// Vitest Snapshot v1

exports[`json body > array input > requires at least one source 1`] = `
{
  "error": "Bad Request",
  "message": "body must be object, body/0 must NOT have fewer than 1 properties, body must match exactly one schema in oneOf",
  "statusCode": 400,
}
`;

exports[`json body > object input > GET fails with json body 1`] = `
{
  "error": "Bad Request",
  "message": "querystring must have required property 'source'",
  "statusCode": 400,
}
`;

exports[`json body > object input > errors correctly on an empty object 1`] = `
{
  "error": "Bad Request",
  "message": "body must NOT have fewer than 1 properties, body must be array, body must match exactly one schema in oneOf",
  "statusCode": 400,
}
`;
